gnc/sub8_controller/nodes/pd_controller.cc

```diff
@@ -0,0 +1,218 @@
+  Eigen::Vector3d current;
+  Eigen::Vector3d previous;
+
+  for (controller_deque::iterator it = history.begin() + 1; it != history.end(); ++it) {
+    current = *it;
+    previous = *(it - 1);
+    sum_difference += (current - previous) / control_period;
+  }
+  gradient = sum_difference / history.size();
+  return gradient;
+}
+
+Eigen::AngleAxis<double> PDController::rotation_difference(Eigen::Matrix3d &rotation_a, Eigen::Matrix3d &rotation_b) {
+  Eigen::AngleAxis<double> angle_axis;
+  double angle;
+  Eigen::Matrix3d difference_matrix;
+  Eigen::Vector3d euler_error;
+
+  difference_matrix = rotation_b * rotation_a.transpose();
+  angle_axis.fromRotationMatrix(difference_matrix);
+  return angle_axis;
+}
+
+void PDController::gain_callback(sub8_controller::GainConfig &config, uint32_t level) {
+  /* Set gains using dynamic reconfigure */
+  kp_trans = config.kp_trans;
+  kd_trans = config.kd_trans;
+  ki_trans = config.ki_trans;
+  kp_angle = config.kp_angle;
+  kd_angle = config.kd_angle;
+  ki_angle = config.ki_angle;
+  trans_history_length = config.history_length;
+  sub_mass = config.sub_mass;
+}
+
+void PDController::truth_callback(const nav_msgs::Odometry::ConstPtr &odom_msg) {
+  /* Callback on truth data */
+  Eigen::Affine3d pose_matrix;
+  Eigen::Matrix<double, 6, 1> twist;
+  Eigen::Vector3d linear_velocity;
+  Eigen::Vector3d angular_velocity;
+
+  got_state = true;
+
+  tf::poseMsgToEigen(odom_msg->pose.pose, pose_matrix);
+  current_state.position = pose_matrix.translation();
+  current_state.orientation = pose_matrix.rotation();
+
+  tf::twistMsgToEigen(odom_msg->twist.twist, twist);
+  current_state.linear_velocity << twist.head(3);
+  current_state.angular_velocity << twist.tail<3>();
+}
+
+void PDController::waypoint_callback(const sub8_msgs::Trajectory::ConstPtr &target_pose_msg) {
+  /* Callback on target waypoint */
+  Eigen::Affine3d target_pose_matrix;
+  Eigen::Matrix<double, 6, 1> twist;
+  Eigen::Vector3d linear_velocity;
+  Eigen::Vector3d angular_velocity;
+
+  got_target = true;
+
+  // Clear the error history
+  translation_error_history.erase(translation_error_history.begin(), translation_error_history.end());
+
+  tf::poseMsgToEigen(target_pose_msg->trajectory[0].pose, target_pose_matrix);
+  target_state.position = target_pose_matrix.translation();
+  target_state.orientation = target_pose_matrix.rotation();
+  if (target_state.orientation.norm() == 0.0) {
+    // If somebody passes a target with no orientation set, do something that makes sense
+    if (got_state) {
+      target_state.orientation = current_state.orientation;
+    } else {
+      target_state.orientation = Eigen::Affine3d::Identity().rotation();
+    }
+  }
+
+  tf::twistMsgToEigen(target_pose_msg->trajectory[0].twist, twist);
+  target_state.linear_velocity << twist.head(3);
+  target_state.angular_velocity << twist.tail<3>();
+}
+
+int main(int argc, char **argv) {
+  ROS_INFO("Starting up controller");
```

Comment:
I ran into some issues before with the log level not automatically being set high enough for ROS_INFO log statements to display. Did you verify that you can see these when running this node?